### ISSUE TITLE
Prediction speed worse than community keras due to training related code running during inference

### ISSUE THREAD

**Issue by drewm1980** [tensorflow/tensorflow#31975]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): YES
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Linux Ubuntu 18.04.2 (but we believe this to be irrelevant)
- TensorFlow installed from (source or binary): pip (inside a docker image)
- TensorFlow version (use command below): 1.14 (likely affects 1.12 - current 2.0 beta)
- Python version: 3.6.8 (system python, we believe irrelevant)
- CUDA/cuDNN version: 10.0.130 (but we believe irrelevant)
- GPU model and memory: 1080 Ti

**Describe the current behavior**
Calling model.predict() includes calls to various non-prediction related things, in order of decreasing severity:
* reset_metrics
* get_progbar
* standardize_user_data
* validate_or_infer_batch_size

In our project, this results in a 2X prediction speed regression:
35 ms per call for community keras
70 ms per call for tensorflow keras

A Snakeviz flamegraph for prediction of our network in Tensorflow Keras, showing unnecessary overhead:

![Tensorflow Keras](https://user-images.githubusercontent.com/190617/63684638-ab884a80-c7fd-11e9-9dbe-5404eda2d70c.png)

**Describe the expected behavior**
Only do things necessary for prediction while doing prediction.

A Snakeviz flamegraph for prediction of our network in Community Keras, showing no unnecessary overhead:

![Community Keras](https://user-images.githubusercontent.com/190617/63684637-ab884a80-c7fd-11e9-810d-6341eba99070.png)

**Code to reproduce the issue**
This code is active for all calls to predict().

This is for a commercial project we are trying to migrate from community keras to tensorflow keras. It is a speed critical real-time robotics application where the 30ms is enough to miss our hard deadlines. If predict() is intended for production inference, we feel this should be resolved in tensorflow keras. If it is only intended for use during training, this should be noted in the mainline documentation, along with a suggestion for what ~is intended for production inference.

As a temporary workaround, we disabled some of these function calls in a local tensorflow fork, and the performance regression went away as expected. We took care to make sure this is not a profiling glitch. We excluded warmup time from the profile, and average over a number of samples.

**Comment by drewm1980** [comment]
Looking at the code in the 2.0 branch, this affects tensorflow 2.0 beta as well. It is not limited to tensorflow 1.14.

We haven't backported our code to 1.12 to test it, but looking at the code, it seems much of the overhead may have been introduced in the commit:

https://github.com/tensorflow/tensorflow/commit/4ca599630233afe5240deae44212f19412ab6423#diff-02db763382946f7695e505ce1c545dc4L250

See the addition at line:
tensorflow/python/keras/engine/training_arrays.py:250

It is possible the issue goes even farther back; we haven't audited the old attribute based mechanism.

**Comment by drewm1980** [comment]
Thanks for looking into this! We can't share our model publicly as-is.

I thought maybe the version of keras-team might have more overhead than the one we benchmarked. The keras-team we were using for benchmark was 2.2.4. We just looked at the current master commit: https://github.com/keras-team/keras/commit/a39f10acf3eac5e6ef69d084616982a4d4ebd8bb and the overall logic has not changed. 

If you look at the predict_loop in keras-team:

https://github.com/keras-team/keras/blob/master/keras/engine/training_arrays.py

You will see that the progress bar code is behind a "verbose" flag that is off by default, and since the prediction is factored in a separate loop from training, the other code called in the tf.keras version doesn't get called either. keras-team does a bunch of batching logic even when there is only a single sample, but apparently it's not as much overhead as the extra logic in the tf.keras version. 

Maybe the extra training logic that was left on for prediction scales badly in the number of tensors? Our network is a weird custom convnet+MLP that probably has more operations than most networks of the same run time, mostly in the MLP part. Maybe the performance issue is reproducible with a really deep MLP with really small layers, but I would not bother unless you're interested in it as an additional stress test. Instead of investing effort in studying why that code is slow in our use case, I would advocate disabling it all behind a flag that is off by default. 

Training and profiling related code should not be active by default in a function that's only supposed to do prediction, right?

**Comment by drewm1980** [comment]
Thanks for looking into this!  We are still following and interested!

**Comment by robieta** [comment]
Sure. I am currently trying to get a fix into the 2.0 release, but let me give an update in the mean time.

In Keras, there are a bunch of properties which recursively depend on all child layers. `metrics` is one, but there are others: `stateful`, `dynamic`, `layers`, etc. What I noticed during profiling is that every `model.predict` call we have to check whether these attributes have changed, and walking that graph was still causing significant python overhead even after I stopped collecting metrics in predict. (~12 ms for the 250 layer DNN that I'm testing on.)

The change that I'm working on adds a layer aware caching mechanism, so we can avoid walking the entire graph-of-layers if nothing has changed. (And generally people aren't going to repeatedly toggle `stateful` or add layers when doing inference...) However, because of the nature of this cache it's vital that we ensure correctness and check that it will invalidate on any potentially relevant mutation, so it's taking a bit longer than I initially anticipated.

I'm really hoping to get this into 2.0. `model.predict` is definitely my recommended way to do inference directly from a keras model since it will make a `tf.function` for you, so I want that path to be a good experience. (There is pure c++ machinery for the really latency sensitive, but I know a lot of folks just want a reasonably convenient way in Python.)
